### ISSUE TITLE
AJAX telemetry has timestamp at end of request rather than start

### ISSUE THREAD

**Issue by OsvaldoRosado** [microsoft/ApplicationInsights-JS#628]
AJAX telemetry appears to record the event timestamp at the time of collection rather than at the time of the call being made. This can be reproduced by setting up a simple server that responds slowly to a request. The implication of this is that the Portal UI can present misleading timelines of what actually happened during the pageview.

For example, I have an AJAX that takes 5 seconds to complete (as shown by Fiddler):
![image](https://user-images.githubusercontent.com/4926351/42234966-13ed1c42-7eab-11e8-8ea2-eb0748479215.png)

When inspecting the generated telemetry envelope, we see the recorded time as after ClientDoneResponse:
![image](https://user-images.githubusercontent.com/4926351/42235028-44d22c30-7eab-11e8-8bf3-dffbc0eeef96.png)

From the [schema description of the time field][0], I would expect the time of my AJAX starting to be the time reported in the envelope, though admittedly it's not clear what exactly "event" refers to:

> Event date time when telemetry item was created. This is the wall clock time on the client when the event was generated. 

It is clearer in the context of these two issues however that it refers to the event the telemetry item is created for, not the instantiation time of the telemetry item itself:
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/658
https://github.com/Microsoft/ApplicationInsights-Home/issues/133

Interestingly, this problem doesn't seem to be limited to autocollection. From what I can tell by a quick look in the code, there doesn't appear to be any way currently to adjust the timestamp of these events, even if manually collecting a dependency via `RemoteDependencyData`.

[0]: https://github.com/Microsoft/ApplicationInsights-Home/blob/master/EndpointSpecs/Schemas/Docs/Envelope.md

**Comment by jpiyali** [comment]
@OsvaldoRosado  - the time field is common for all envelopes (not specific to dependency item). It is not clear from the description that it is meant to be the contextual (refers to the event the telemetry item is created for). It would make more sense to keep this time to the time the event was sent (which in most cases is time when event is instantiated to be sent). For things like dependency start/end time, there should be fields specific to domain data (part B) for this. @SergeyKanzhelev  - how is the time field supposed to be treated?

**Comment by OsvaldoRosado** [comment]
This seems to be answered by @SergeyKanzhelev explicitly in https://github.com/Microsoft/ApplicationInsights-Home/issues/133#issuecomment-314651906 if that's helpful.

Quoting for convenience:
>timestamp represents the time event was "created". For request and dependency telemetry it will be a start time of said request and dependencies (time when you called StartOperation method of telemetry client). For trace and event - time telemetry item was tracked (Track method of telemetry client was called).

**Comment by vdesmedt** [comment]
Extremly interrested into this as we need to track the time between first byte received and first tracking dependency fired ! At the moment, we suppose that the timestamp is from the moment the telemetry is marked as ready for send and subtract the duration from it to infer the starting time but that gives less than convincing results.

**Comment by paulyoder** [comment]
Unfortunately this looks like it won't be an easy fix.

The `AjaxMonitor` class is responsible for recording the Ajax telemetry as a dependency event. You can see from its `onAjaxComplete` method below ([link](https://github.com/Microsoft/ApplicationInsights-JS/blob/master/JavaScript/JavaScriptSDK/ajax/ajax.ts#L256)) that it creates a `RemoteDependencyData` object and passes it to `appInsight.trackDependencyData()`

```javascript
private onAjaxComplete(xhr: XMLHttpRequestInstrumented) {
  ...
  var dependency = new Telemetry.RemoteDependencyData(
    xhr.ajaxData.id,
    ...
  );
  ...
  this.appInsights.trackDependencyData(dependency);
}
```

Then `trackDependencyData` creates an `Envelope` object that wraps the dependency data and sends it on to the server.  ([link](https://github.com/Microsoft/ApplicationInsights-JS/blob/27f4f30ff06e58a8095ea847eba258de263256c3/JavaScript/JavaScriptSDK/AppInsights.ts#L323))

```javascript
public trackDependencyData(dependency: Telemetry.RemoteDependencyData) {
  ...
  var dependencyData = new ApplicationInsights.Telemetry.Common.Data<ApplicationInsights.Telemetry.RemoteDependencyData>(
                    Telemetry.RemoteDependencyData.dataType, dependency);
  var envelope = new Telemetry.Common.Envelope(dependencyData, ApplicationInsights.Telemetry.RemoteDependencyData.envelopeType);
  this.context.track(envelope);
  ...
}
```

The timestamp gets set when the `Envelope` object is created in `appInsights.trackDependencyData()` which is called only after the Ajax response completes ([link](https://github.com/Microsoft/ApplicationInsights-JS/blob/8e9701d737a5391328d86f1fecba66ef4b2b1396/JavaScript/JavaScriptSDK/Telemetry/Common/Envelope.ts#L22)). There is no current way to override the timestamp when calling `trackDependencyData()`, so it's not possible for the `AjaxMonitor` to override the timestamp.

```javascript
export class Envelope extends Microsoft.Telemetry.Envelope implements IEnvelope {
  constructor(data: Microsoft.Telemetry.Base, name: string) {
    ...
    this.time = Util.toISOStringForIE8(new Date());
    ...
  }
}
```

It would be nice to have the option to manually set the timestamp when calling `appInsights.trackDependencyData()`. Hopefully I'm missing something and there is a way to manually set the timestamp from the `AjaxMonitor` class.

**Comment by OsvaldoRosado** [comment]
Could we fix this by adding a date field to the `Telemetry.RemoteDependencyData` interface, and then overriding the time property of the envelope with its value when set?


This is essentially the approach used in the Node.js SDK: https://github.com/Microsoft/ApplicationInsights-node.js/blob/develop/Library/TelemetryClient.ts#L126
```
    public track(telemetry: Contracts.Telemetry, telemetryType: Contracts.TelemetryType) {
        if (telemetry && Contracts.telemetryTypeToBaseType(telemetryType)) {
            var envelope = EnvelopeFactory.createEnvelope(telemetry, telemetryType, this.commonProperties, this.context, this.config);

            // Set time on the envelope if it was set on the telemetry item
            if (telemetry.time) {
                envelope.time = telemetry.time.toISOString();
            }
...
```

**Comment by MSNev** [comment]
I believe that this was addressed in v2.6.0
